ESP32: Add bridge app example to ESP32

sweetymhaiske opened the pull request:
#### Problem
* Add bridge-app example to ESP32.

#### Change overview
* Created 4 devices (non-chip devices).
* Created dynamic endpoints, clusters, and attributes
* Non CHIP devices are added to the endpoint 2, 4, 5, 6.

#### Testing
* Build and flash the application on esp32 DevkitC
* Performed on-off cluster control using python chip controller on the dynamically created endpoints.

**command 1**
```
chip-device-ctrl > zclread Descriptor PartsList 135246 0 0
```
**output 1**
```
[1626447466.983630][446912:446920] CHIP:ZCL: ReadAttributesResponse:
[1626447466.983634][446912:446920] CHIP:ZCL:   ClusterId: 0x001d
[1626447466.983640][446912:446920] CHIP:ZCL:   attributeId: 0x0003
[1626447466.983642][446912:446920] CHIP:ZCL:   status: Success                (0x0000)
[1626447466.983649][446912:446920] CHIP:ZCL:   attribute TLV Type: 0x17
[1626447466.983659][446912:446920] CHIP:ZCL:   attributeValue: List of length 4
```
**command 2**
```
chip-device-ctrl > zcl OnOff Off 135246 4 0
```
**output 2**
```
I (217115) chip[ZCL]: On/Off set value: 4 0
I (217125) chip[DL]: HandleReadOnOffAttribute: attrId=0, maxReadLength=1
I (217125) chip[ZCL]: Toggle on/off from 1 to 0
I (217135) chip[DL]: HandleWriteOnOffAttribute: attrId=0
I (217145) chip[DL]: Device[Light 3]: OFF

```

andy31415 commented on examples/bridge-app/esp32/main/Device.cpp:
```diff
@@ -0,0 +1,118 @@
+        ChipLogProgress(DeviceLayer, "Device[%s]: OFF", mName);
+    }
+
+    if ((changed) && (mChanged_CB))
```
could we remove some of the extra bracketing here and below?

andy31415 commented on examples/bridge-app/esp32/main/DeviceCallbacks.cpp:
```diff
@@ -0,0 +1,125 @@
+/**
+ * @file DeviceCallbacks.cpp
+ *
+ * Implements all the callbacks to the application from the CHIP Stack
```
consider deleting this comment: it seems obvious that `DeviceCallbacks` implements callbacks.

andy31415 commented on examples/bridge-app/esp32/main/DeviceCallbacks.cpp:
```diff
@@ -0,0 +1,125 @@
+    VerifyOrExit(attributeId == ZCL_ON_OFF_ATTRIBUTE_ID, ESP_LOGI(TAG, "Unhandled Attribute ID: '0x%04x", attributeId));
+    VerifyOrExit(endpointId == 1 || endpointId == 2, ESP_LOGE(TAG, "Unexpected EndPoint ID: `0x%02x'", endpointId));
+    /*Actions on fixed endpoints*/
+exit:
```
please use return macros instead of exit macros, since this exit does not contain any cleanup code.

it may be more readable to just use if considitons and no macros at all:

```cpp

if (attributeId != ZCL_ON_OFF_ATTRIBUTE_ID) {
  // is logI correct here, or should it be an error?
  ESP_LOGI(TAG, "Unhandled Attribute ID: '0x%04x", attributeId);
  return;
}

if ((endpointId < 1) || (endpointId > 2)) {
  ESP_LOGE(TAG, "Unexpected EndPoint ID: `0x%02x'", endpointId);
  return;
}

sweetymhaiske replied:
Done

andy31415 commented on examples/bridge-app/esp32/main/include/Device.h:
```diff
@@ -0,0 +1,67 @@
+
+    Device(const char * szDeviceName, const char * szLocation);
+
+    bool IsOn();
```
I think the getters can be `const`

andy31415 commented on examples/bridge-app/esp32/main/main.cpp:
```diff
@@ -0,0 +1,452 @@
+        {
+            gDevices[index] = dev;
+            EmberAfStatus ret;
+            while (1)
```
what guarantees that this while loop returns? I.e. why do we know that emberAfSetDynamicEndpoint will eventually eitehr succeed or return 'duplicate exists'?

Some comments would be nice or alternatively a fixed size loop would also help.

sweetymhaiske replied:
Removed infinite loop

andy31415 commented on examples/bridge-app/esp32/main/main.cpp:
```diff
@@ -0,0 +1,452 @@
+{
+    ChipLogProgress(DeviceLayer, "HandleReadOnOffAttribute: attrId=%d, maxReadLength=%d", attributeId, maxReadLength);
+
+    if ((attributeId == ZCL_ON_OFF_ATTRIBUTE_ID) && (maxReadLength == 1))
```
nit: if you switch the if condition, there may be less nesting:

```cpp
ReturnErrorCodeIf(
    (attributeId != ZCL_ON_OFF_ATTRIBUTE_ID) || (maxReadLength != 1), 
    EMBER_ZCL_STATUS_FAILURE);

*buffer = dev->IsOn() ? 1 : 0;
return EMBER_ZCL_STATUS_SUCCESS;
```

sweetymhaiske replied:
done

andy31415 commented on examples/bridge-app/esp32/main/main.cpp:
```diff
@@ -0,0 +1,452 @@
+    {
+        if (*buffer)
+        {
+            dev->SetOnOff(true);
```
`dev->SetOnOff(*buffer !=  0);` could be shorter. I would also invert the condition to do an early 'return error' to reduce nesting. Code that reads `if (cond) { return ERR; }....`  seems easier to parse than having the return error at the tail.

```cpp
if (attributeId != ZCL_ON_OFF_ATTRIBUTE_ID) || !dev->IsReachable()) {
  return EMBER_ZCL_STATUS_FAILURE;
}

dev->SetOnOff(*buffer !=  0);
return EMBER_ZCL_STATUS_SUCCESS;
```

or you could use return macros for even shorter code.

sweetymhaiske replied:
done

andy31415 commented on examples/bridge-app/esp32/main/main.cpp:
```diff
@@ -0,0 +1,452 @@
+{
+    uint16_t endpointIndex = emberAfGetDynamicIndexFromEndpoint(endpoint);
+
+    EmberAfStatus ret = EMBER_ZCL_STATUS_FAILURE;
```
do we need to declare the ret?

how about directly returning in the if statements below?

andy31415 replied:
Same comment for the methods below

andy31415 commented on examples/bridge-app/esp32/main/main.cpp:
```diff
@@ -0,0 +1,452 @@
+
+    EmberAfStatus ret = EMBER_ZCL_STATUS_FAILURE;
+
+    // ChipLogProgress(DeviceLayer, "emberAfExternalAttributeWriteCallback: ep=%d", endpoint);
```
remove commented out code

bzbarsky-apple commented on examples/bridge-app/esp32/main/main.cpp:
```diff
@@ -0,0 +1,452 @@
+
+    labelStruct.label = chip::ByteSpan(reinterpret_cast<const uint8_t *>(label), kFixedLabelElementsOctetStringSize);
+
+    strncpy(zclOctetStrBuf, value, sizeof(zclOctetStrBuf));
```
`CopyString`, please.  But honestly, can we not just use `(value, strlen(value))` as a Span without having to do this copy?

sweetymhaiske replied:
Yup. Did it.

bzbarsky-apple commented on examples/bridge-app/esp32/main/main.cpp:
```diff
@@ -0,0 +1,452 @@
+    uint16_t listCount = 1;
+    _LabelStruct labelStruct;
+
+    labelStruct.label = chip::ByteSpan(reinterpret_cast<const uint8_t *>(label), kFixedLabelElementsOctetStringSize);
```
This does not make any sense.  Callers pass things like "room", but here we are going to read `kFixedLabelElementsOctetStringSize == 16` bytes from that pointer?  That's going to read random memory...  

I really wish we were adding tests to go along with this code....

sweetymhaiske replied:
Yeah. Corrected.

dhrishi commented on examples/bridge-app/esp32/main/CMakeLists.txt:
```diff
@@ -0,0 +1,40 @@
+                      "${CMAKE_SOURCE_DIR}/third_party/connectedhomeip/src/app/clusters/on-off-server"
+                      "${CMAKE_SOURCE_DIR}/third_party/connectedhomeip/src/app/clusters/operational-credentials-server"
+                      "${CMAKE_SOURCE_DIR}/third_party/connectedhomeip/src/app/clusters/general-commissioning-server"
+                      PRIV_REQUIRES chip QRCode tft spidriver bt screen-framework)
```
tft, spidriver, screen-framework not required

dhrishi commented on examples/bridge-app/esp32/main/CMakeLists.txt:
```diff
@@ -0,0 +1,40 @@
+                      "${CMAKE_SOURCE_DIR}/third_party/connectedhomeip/src/app/clusters/general-commissioning-server"
+                      PRIV_REQUIRES chip QRCode tft spidriver bt screen-framework)
+
+set_property(TARGET ${COMPONENT_LIB} PROPERTY CXX_STANDARD 17) 
```
Same question here.. is C++17 required?

dhrishi commented on examples/bridge-app/esp32/main/Kconfig.projbuild:
```diff
@@ -0,0 +1,44 @@
+          bool "Ethernet"
+    endchoice
+
+    config USE_ECHO_CLIENT
```
Is this option used in the example?

dhrishi commented on examples/bridge-app/esp32/main/Kconfig.projbuild:
```diff
@@ -0,0 +1,44 @@
+            This enables a local FreeRTOS Echo Client so that the end-to-end echo server can be
+            tested easily
+
+    config ECHO_HOST_IP
```
Is this option used in the example?

dhrishi commented on examples/bridge-app/esp32/main/main.cpp:
```diff
@@ -0,0 +1,452 @@
+    int err = 0;
+    // Initialize the ESP NVS layer.
+    err = nvs_flash_init();
+    if (err != CHIP_NO_ERROR)
```
nvs_flash_init returns esp_err_t. So you should check the return value with `ESP_OK`

dhrishi commented on examples/bridge-app/esp32/main/main.cpp:
```diff
@@ -0,0 +1,452 @@
+    AddDeviceEndpoint(&Light4, &bridgedLightEndpoint, DEVICE_TYPE_LO_ON_OFF_LIGHT);
+
+    // Re-add Light 2 -- > will be mapped to ZCL endpoint 6
+    AddDeviceEndpoint(&Light2, &bridgedLightEndpoint, DEVICE_TYPE_LO_ON_OFF_LIGHT);
```
If majority of the code here is exactly same for the Linux bridge example, would it make sense to move it to some common location, may be `examples/bridge-app/bridge-common/`. This is just an observation and I am not sure about this. Can you file an issue so that it can be discussed?

sweetymhaiske replied:
I think this should be example-specific. ESP32 was written, with reference to Linux, hence they are the same, this use-case may or may not be the same for other platforms. However I have created the issue. #8458

dhrishi replied:
Yes, that's right, some part of it would be example specific. But, still APIs like `AddDeviceEndpoint`, `RemoveDeviceEndpoint`, different callbacks and utility functions can be made common.

dhrishi commented:
@sweetymhaiske Please add the actual controller commands in the Testing section.

sweetymhaiske commented on examples/bridge-app/esp32/main/main.cpp:
```diff
@@ -0,0 +1,452 @@
+    emberAfEndpointEnableDisable(emberAfEndpointFromIndex(static_cast<uint16_t>(emberAfFixedEndpointCount() - 1)), false);
+
+    // Add lights 1..3 --> will be mapped to ZCL endpoints 2, 3, 4
+    AddDeviceEndpoint(&Light1, &bridgedLightEndpoint, DEVICE_TYPE_LO_ON_OFF_LIGHT);
```
Done

sweetymhaiske commented on examples/bridge-app/esp32/main/CHIPDeviceManager.cpp:
```diff
@@ -0,0 +1,104 @@
+
+    // Initialize the CHIP stack.
+    err = PlatformMgr().InitChipStack();
+    SuccessOrExit(err);
```
Done.

sweetymhaiske commented on examples/bridge-app/esp32/main/Device.cpp:
```diff
@@ -0,0 +1,118 @@
+{
+    bool changed = (strncmp(mLocation, szLocation, sizeof(mLocation)) != 0);
+
+    strncpy(mLocation, szLocation, sizeof(mLocation));
```
Done

sweetymhaiske commented on examples/bridge-app/esp32/CMakeLists.txt:
```diff
@@ -0,0 +1,30 @@
+)
+
+project(chip-bridge-app)
+idf_build_set_property(CXX_COMPILE_OPTIONS "-std=gnu++17;-Os;-DLWIP_IPV6_SCOPES=0;-DCHIP_HAVE_CONFIG_H;-DDYNAMIC_ENDPOINT_COUNT=16" APPEND)
```
Nothing specific. Changed to C++14.

sweetymhaiske commented on examples/bridge-app/esp32/main/main.cpp:
```diff
@@ -0,0 +1,452 @@
+static uint8_t gDescriptorAttrStorage[DYNAMIC_ENDPOINT_COUNT][kDescriptorAttributeCount][kDescriptorAttributeArraySize];
+
+// (taken from chip-devices.xml)
+#define DEVICE_TYPE_CHIP_BRIDGE 0x0a0b
```
Removed.

sweetymhaiske replied:
I have not found this device type being exposed.

bzbarsky-apple replied:
Right, my point is that we should fix that. Hence asking for an issue for that.

sweetymhaiske replied:
https://github.com/project-chip/connectedhomeip/issues/2764

sweetymhaiske commented on examples/bridge-app/esp32/main/main.cpp:
```diff
@@ -0,0 +1,452 @@
+    uint16_t listCount = 1;
+    _LabelStruct labelStruct;
+
+    labelStruct.label = chip::ByteSpan(reinterpret_cast<const uint8_t *>(label), kFixedLabelElementsOctetStringSize);
```
Done

sweetymhaiske commented:
> @sweetymhaiske Please add the actual controller commands in the Testing section.

Added.

sweetymhaiske commented on examples/bridge-app/esp32/main/Device.cpp:
```diff
@@ -0,0 +1,118 @@
+
+Device::Device(const char * szDeviceName, const char * szLocation)
+{
+    strncpy(mName, szDeviceName, sizeof(mName));
```
Done

sweetymhaiske commented:
@dhrishi @bzbarsky-apple @andy31415  PTAL.

sweetymhaiske commented:
@bzbarsky-apple PTAL

bzbarsky-apple reviewed (commented):
Thank you for addressing most of the issues!

The two problems remaining are:

1) This is still adding yet another copy/paste of `ToZclCharString` instead of putting it as a utility somewhere shared.  I'd really prefer for this to be fixed before merging this, but can probably live with a commitment to do it immediately after this merges.
2) No tests.  Followup for this, I guess, since testing on-device things is not trivial.

sweetymhaiske commented:
> Thank you for addressing most of the issues!
> 
> The two problems remaining are:
> 
> 1. This is still adding yet another copy/paste of `ToZclCharString` instead of putting it as a utility somewhere shared.  I'd really prefer for this to be fixed before merging this, but can probably live with a commitment to do it immediately after this merges.
> 2. No tests.  Followup for this, I guess, since testing on-device things is not trivial.

1. I am waiting for this PR #8325 to get merged. Once done I assure you, I will add those changes in this app too.
2. For now, I have added the local test using python controller in PR description.

sweetymhaiske commented:
@bzbarsky-apple  PTAL

bzbarsky-apple commented:
@andy31415 are your comments here addressed?

sweetymhaiske commented:
@andy31415 Please help to get this merge.

bzbarsky-apple commented:
@sweetymhaiske Should this new app have a README.md explaining how to build and whatnot?

bzbarsky-apple commented:
@sweetymhaiske And shouldn't `examples/bridge-app/esp32/build/` and `examples/bridge-app/esp32/sdkconfig` be added to `.gitignore`?
